File: reinforcementNeuralNetwork/game/game.py
import pygame
import random
import time
import socket

SOCKET_PORT = 1234
BUFFER_SIZE = 1024

# Initialize Pygame
pygame.init()

# Constants
WIDTH, HEIGHT = None, None # none for now and will be updated when the game starts
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
FPS = 120#360
MOVE_FILE_PATH = "../data/move.txt"
NETWORK_INPUT_FILE_PATH = "../data/gameStateData.txt"

# enable or disable the ability to view the game
viewGame = 0

# enable or disable keyboard input for user controlled paddle (controls left paddle)
userControlledPaddle = True

# enable or disable single player mode
singlePlayerMode = False

# initialize socket
clientSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
clientSocket.connect(('127.0.0.1', SOCKET_PORT))
print("connected to server")

def recvGameConstants(clientSocket):
    # Receive the game constants from the server
    data = clientSocket.recv(BUFFER_SIZE).decode('utf-8')
    print(f"Received game constants: {data}")
    # Parse the game constants (width, height)
    width, height = map(float, data.strip("{}").split(","))
    return width, height

def recvMove(clientSocket):
    # recieve move from the server
    data = clientSocket.recv(BUFFER_SIZE).decode('utf-8')
    # parse the move
    move = float(data.strip("{}"))
    return move

def sendGameState(client_socket, game_state):
    # Send the game state to the server
    client_socket.send(game_state.encode('utf-8'))

# read the game constants
while WIDTH is None or HEIGHT is None:
    WIDTH, HEIGHT = recvGameConstants(clientSocket)
    print(f"Game Constants - Width: {WIDTH}, Height: {HEIGHT}")

# player points
opponentPoints = 0
playerPoints = 0

# define the current player direction
# 0.5: no movement, greater than 0.5: up, less than 0.5: down
playerDirection = None

# same as above but previous direction (init to none for the first iteration)
prevPlayerDirection = None

# Create the screen
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Pong")
clock = pygame.time.Clock()

# Define the game objects
paddleWidth = 10
paddleHeight = 100
ballSize = 10

# Paddle initialization
if not singlePlayerMode:
    leftPaddle = pygame.Rect(50, HEIGHT // 2 - paddleHeight // 2, paddleWidth, paddleHeight)
else:
    # fill entire left side of the screen with a paddle
    leftPaddle = pygame.Rect(0, 0, paddleWidth, HEIGHT)
rightPaddle = pygame.Rect(WIDTH - 50 - paddleWidth, HEIGHT // 2 - paddleHeight // 2, paddleWidth, paddleHeight)

# Ball initialization
ball = pygame.Rect(WIDTH // 2 - ballSize // 2, HEIGHT // 2 - ballSize // 2, ballSize, ballSize)
ballVelocityX = random.choice([-3, 3])
ballVelocityY = random.choice([-3, 3])

# Send the initial game state
#gameState = "{" + str(rightPaddle.y) + "," + str(leftPaddle.y) + "," + str(playerPoints) + "," + str(opponentPoints) + "," + str(ball.x) + "," + str(ball.y) + "," + str(ballVelocityX) + "," + str(ballVelocityY) + "}"
#sendGameState(clientSocket, gameState)

# Main game loop
running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    # Ball movement
    ball.x += ballVelocityX
    ball.y += ballVelocityY

    # Ball collision with paddles (handle collisions with sides and top)
    # if the ball hits the top or bottom of the paddles, make it go a bit faster in the y axis. if it hits the middle it should go forward straight
    if ball.colliderect(leftPaddle) and not singlePlayerMode:
        # Reverse horizontal velocity and adjust vertical velocity based on the position of the ball on the paddle
        ballVelocityX = abs(ballVelocityX)  # Ensure the ball always moves to the right after collision
        ballVelocityY = int((ball.centery - leftPaddle.centery) / (paddleHeight / 2) * 5)  # Adjust vertical velocity between -5 and 5
    elif ball.colliderect(rightPaddle):
        # Reverse horizontal velocity and adjust vertical velocity based on the position of the ball on the paddle
        ballVelocityX = -abs(ballVelocityX)  # Ensure the ball always moves to the left after collision
        ballVelocityY = int((ball.centery - rightPaddle.centery) / (paddleHeight / 2) * 5)  # Adjust vertical velocity between -5 and 5
    elif singlePlayerMode and ball.colliderect(leftPaddle):
        ballVelocityX *= -1
    

    # Ball collision with walls
    if ball.top <= 0 or ball.bottom >= HEIGHT:
        ballVelocityY *= -1

    # Ball out of bounds
    if ball.left <= 0 and not singlePlayerMode:
        # player scores
        playerPoints += 1
        ball.center = (WIDTH // 2, HEIGHT // 2)
        ballVelocityX = random.choice([-3, 3])
        ballVelocityY = random.choice([-3, 3])
    elif ball.right >= WIDTH:
        # opponent scores
        opponentPoints += 1
        ball.center = (WIDTH // 2, HEIGHT // 2)
        ballVelocityX = random.choice([-3, 3])
        ballVelocityY = random.choice([-3, 3])


    # send the current game state
    # right paddle is the player paddle
    # the format: {playerPaddleY,opponentPaddleY,playerPoints,opponentPoints,ballX,ballY,ballVelocityX,ballVelocityY}
    gameState = "{" + str(rightPaddle.y) + "," + str(leftPaddle.y) + "," + str(playerPoints) + "," + str(opponentPoints) + "," + str(ball.x) + "," + str(ball.y) + "," + str(ballVelocityX) + "," + str(ballVelocityY) + "}"
    sendGameState(clientSocket, gameState)
    
    # Right paddle movement (player)
    # Take move input from the server and update the playerDirection variable
    # the format is {moveDirection} so we need to remove the curly braces and read the value
    # loop until the playerDirection is not None and it contains a float value
    playerDirection = recvMove(clientSocket)

    while playerDirection is None:
        try:
            playerDirection = recvMove(clientSocket)
        except:
            playerDirection = None


    # Player/agent (right) movement
    # before moving, make sure the player would be able to move the paddle in the next iteration
    # if the playerDirection is greater than 0.5, move the paddle up
    if playerDirection > 0.5 and rightPaddle.y > 0:
        rightPaddle.y -= 5
    # if the playerDirection is less than 0.5, move the paddle down
    elif playerDirection < 0.5 and rightPaddle.y < HEIGHT - paddleHeight:
        rightPaddle.y += 5
    # update the prevPlayerDirection variable
    prevPlayerDirection = playerDirection


    # Left paddle movement (opponent)
    # if the ball is above the paddle's center, move the paddle up
    if userControlledPaddle:
        # opponent direction controlled by user
        # get input from keyboard
        keys = pygame.key.get_pressed()
        if keys[pygame.K_UP] and leftPaddle.y > 0:
            leftPaddle.y -= 5
        if keys[pygame.K_DOWN] and leftPaddle.y < HEIGHT - paddleHeight:
            leftPaddle.y += 5
    elif not singlePlayerMode:
        if leftPaddle.y + paddleHeight // 2 > ball.y: # + 20:
            leftPaddle.y -= 3
        # if the ball is below the paddle's center, move the paddle down
        elif leftPaddle.y + paddleHeight // 2 < ball.y: # - 20:
            leftPaddle.y += 3

    # prevent paddles from going out of bounds
    # also prevent the left paddle from going right to the edge so we dont have the ball constantly bouncing back and forth
    if rightPaddle.y <= 0:
        rightPaddle.y = 0
    if rightPaddle.y >= HEIGHT - paddleHeight:
        rightPaddle.y = HEIGHT - paddleHeight
    if leftPaddle.y <= 10:
        leftPaddle.y = 10
    if leftPaddle.y >= HEIGHT - paddleHeight - 10:
        leftPaddle.y = HEIGHT - paddleHeight - 10

    # check render.txt file to check if the game should render
    with open("render.txt", "r") as file:
        content = file.read().strip()
        viewGame = content == "1"
        

    # Render
    if viewGame:
        screen.fill(BLACK)
        pygame.draw.rect(screen, WHITE, leftPaddle)
        pygame.draw.rect(screen, WHITE, rightPaddle)
        pygame.draw.ellipse(screen, WHITE, ball)
        pygame.draw.aaline(screen, WHITE, (WIDTH // 2, 0), (WIDTH // 2, HEIGHT))

        font = pygame.font.SysFont(None, 50)
        p2_text = font.render(str(opponentPoints), True, WHITE)
        p1_text = font.render(str(playerPoints), True, WHITE)
        screen.blit(p2_text, (WIDTH // 4, 50))
        screen.blit(p1_text, (WIDTH * 3 // 4, 50))

        pygame.display.flip()
        clock.tick(FPS)

clientSocket.close()
pygame.quit()

File: reinforcementNeuralNetwork/game/game.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <SDL2/SDL.h>

#define PORT1 1234
#define PORT2 1235

#define PADDLE_WIDTH 10
#define PADDLE_HEIGHT 100
#define BALL_SIZE 10

#define USER_CONTROLLED_PADDLE 0
#define DUELING_AGENTS 1
#define FPS 120

int width = 0;
int height = 0;

typedef struct Ball {
    // allow ball vars and SDL_FRect vars to share the same memory space
    union {
        struct {float x, y, w, h;};
        SDL_FRect rect;
    };
    float velX, velY;
} Ball;

typedef struct Paddle {
    // allow paddle vars and SDL_FRect vars to share the same memory space
    union {
        struct {float x, y, w, h;};
        SDL_FRect rect;
    };
} Paddle;

/*
 ** Description
 * Creates a window to be displayed on the screen
 ** Parameters
 * windowTitle: the title that is displayed on the menu bar of the window
 * windowPosX: the x position on the screen to display the window
 * windowPosY: the y position on the screen to display the window
 * windowSizeX: the x size of the displayed window
 * windowSizeY: the y size of the displayed window
 * flags: additional flags for things like look of window, functionality, etc...
 ** Return Value
 * window: returns a pointer to the window that has been displayed or null if error
 */
SDL_Window *createWindow(const char *windowTitle, int windowPosX, int windowPosY, int windowSizeX, int windowSizeY, int flags) {
    if (SDL_Init(SDL_INIT_EVERYTHING) < 0) {
        fprintf(stderr, "Failed to initialize the SDL2 library\n");
        return NULL;
    }

    SDL_Window *window = SDL_CreateWindow(windowTitle, windowPosX, windowPosY, windowSizeX, windowSizeY, flags);

    if (!window) {
        fprintf(stderr, "Failed to create window\n");
        return NULL;
    }

    return window;
}

int connectToServer(int port) {
    int clientSock, status;
    if ((clientSock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        printf("Socket creation error\n");
        return -1;
    }

    struct sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(port);

    if (inet_pton(AF_INET, "127.0.0.1", &serverAddr.sin_addr) <= 0) {
        printf("Invalid address/ Address not supported\n");
        return -1;
    }

    if ((status = connect(clientSock, (struct sockaddr*)&serverAddr, sizeof(serverAddr))) < 0) {
        printf("\nConnection Failed \n");
        return -1;
    }

    return clientSock;
}

int recvGameConstants(int *width, int *height, int clientSock) {
    char constantBuff[50];
    int retVal;
    if ((retVal = recv(clientSock, constantBuff, sizeof(constantBuff)-1, 0)) <= 0) {
        printf("Error receiving game constants\n");
        return -1;
    }
    constantBuff[retVal] = '\0';


    if ((retVal = sscanf(constantBuff, "{%d, %d}", width, height)) != 2) {
        printf("Could not extract game constants\n");
        return -1;
    }

    return 0;
}

void initGameObjs(Ball *ball, Paddle *leftPaddle, Paddle *rightPaddle) {
    // ball
    ball->x = width/2;
    ball->y = height/2;
    ball->w = BALL_SIZE;
    ball->h = BALL_SIZE;
    ball->velX = (rand() % 2 == 0) ? 3 : -3;
    ball->velY = (rand() % 2 == 0) ? 3 : -3;

    // left paddle
    leftPaddle->x = 50;
    leftPaddle->y = height/2 - PADDLE_HEIGHT/2;
    leftPaddle->w = PADDLE_WIDTH;
    leftPaddle->h = PADDLE_HEIGHT;

    // right paddle
    rightPaddle->x = width-50-PADDLE_WIDTH;
    rightPaddle->y = height/2 - PADDLE_HEIGHT/2;
    rightPaddle->w = PADDLE_WIDTH;
    rightPaddle->h = PADDLE_HEIGHT;
}

// returns 1 on collision, 0 otherwise
int collideRect(SDL_FRect *rect1, SDL_FRect *rect2) {
    // rect1 inside rect2 (left, right, top, bottom impacts respectively)
    if (rect1->x + rect1->w > rect2->x && 
        rect1->x < rect2->x + rect2->w && 
        rect1->y + rect1->h > rect2->y && 
        rect1->y < rect2->y + rect2->h
        ) return 1;

    return 0;
}

int main() {
    int renderGame = 0;
    int leftPts = 0, rightPts = 0;

    int clientSock1 = connectToServer(PORT1);
    if (clientSock1 < 0) {
        return -1;
    }

    int clientSock2 = connectToServer(PORT2);
    if (clientSock2 < 0) {
        return -1;
    }

    int retVal = recvGameConstants(&width, &height, clientSock1);
    if (retVal < 0) {
        return -1;
    }

    // not really needed. just allows the second client to continue the process
    retVal = recvGameConstants(&width, &height, clientSock2);
    if (retVal < 0) {
        return -1;
    }

    printf("game constants: %dx%d\n", width, height);

    // create window and renderer
    SDL_Window *window = createWindow("pong", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, width, height, 0);
    SDL_Renderer* renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);

    srand(time(NULL)); // init random number stuff


    int gameRunning = 1;

    // init game objs
    Ball ball;
    Paddle leftPaddle;
    Paddle rightPaddle;
    initGameObjs(&ball, &leftPaddle, &rightPaddle);

    // init background stuff
    SDL_FRect line;
    line.x = width/2;
    line.y = 0;
    line.w = 3;
    line.h = height;

    // get pointer to internal auto-updating key state array
    const Uint8 *keyState = SDL_GetKeyboardState(NULL);

    int framesSinceFileRead = 0;

    while (gameRunning) {
        SDL_Event e;
        while(SDL_PollEvent(&e) > 0) {
            if (e.type == SDL_QUIT) {
                gameRunning = 0;
            }
        }

        //
        // Game Updates
        //

        ball.x += ball.velX;
        ball.y += ball.velY;

        // ball collision with paddles
        if (collideRect(&ball.rect, &leftPaddle.rect)) {
            // reverse horizontal velocity and adjust vertical velocity based on the position of the ball on the paddle
            ball.velX = fabs(ball.velX); // force ball to go right
            ball.velY = ((ball.y+(ball.h/2)) - (leftPaddle.y+(leftPaddle.h/2))) / (PADDLE_HEIGHT/2) * 5; // adjust vertical velocity between -5 and 5
        } else if (collideRect(&ball.rect, &rightPaddle.rect)) {
            ball.velX = -fabs(ball.velX); // force ball to go left
            ball.velY = ((ball.y+(ball.h/2)) - (rightPaddle.y+(rightPaddle.h/2))) / (PADDLE_HEIGHT/2) * 5; // adjust vertical velocity between -5 and 5
        }

        // ball collision with walls
        if (ball.y <= 0 || ball.y+ball.h >= height) {
            ball.velY *= -1;
        }

        // ball out of bounds
        if (ball.x <= 0) {
            // right scores & reset
            rightPts++;
            ball.x = width/2;
            ball.y = width/2;
            ball.velX = (rand() % 2 == 0) ? 3 : -3;
            ball.velY = (rand() % 2 == 0) ? 3 : -3;
        } else if (ball.x >= width) {
            // left scores & reset
            leftPts++;
            ball.x = width/2;
            ball.y = width/2;
            ball.velX = (rand() % 2 == 0) ? 3 : -3;
            ball.velY = (rand() % 2 == 0) ? 3 : -3;
        }

        // broadcast game state
        // right paddle is the player paddle. left paddle is the opponent paddle
        // the format: {rightPaddleY,leftPaddleY,playerPoints,opponentPoints,ballX,ballY,ballVelocityX,ballVelocityY}
        char gameState[100];
        int len = snprintf(gameState, sizeof(gameState), "{%f,%f,%d,%d,%f,%f,%f,%f}", rightPaddle.y, leftPaddle.y, rightPts, leftPts, ball.x, ball.y, ball.velX, ball.velY);
        if (len >= sizeof(gameState)) {
            send(clientSock1, NULL, 0, 0); // Send nothing. I don't wanna handle cases like this rn
            send(clientSock2, NULL, 0, 0);
        } else {
            send(clientSock1, gameState, len, 0);
            send(clientSock2, gameState, len, 0);
        }

        // recieve right paddle (player) direction
        char playerDirectionBuf[11];
        float playerDirection = -1;
        int bytesRead = 0;
        do {
            bytesRead = recv(clientSock1, playerDirectionBuf, sizeof(playerDirectionBuf)-1, 0);
            if (bytesRead > 0) {
                playerDirectionBuf[10] = '\0';
                sscanf(playerDirectionBuf, "{%f}", &playerDirection);
            } else {
                printf("Error receiving player direction\n");
                gameRunning = 0;
            }
        } while (bytesRead <= 0 && gameRunning);

        // right paddle movement
        if (playerDirection > 0.5f && rightPaddle.y > 0) {
            rightPaddle.y -= 5;
        } else if (playerDirection < 0.5f && rightPaddle.y < height - PADDLE_HEIGHT) {
            rightPaddle.y += 5;
        }

        // left paddle movement
        if (DUELING_AGENTS) {
            // recieve left paddle (opponent) direction
            char opponentDirectionBuf[11];
            float opponentDirection = -1;
            bytesRead = 0;
            do {
                bytesRead = recv(clientSock2, opponentDirectionBuf, sizeof(opponentDirectionBuf)-1, 0);
                if (bytesRead > 0) {
                    opponentDirectionBuf[10] = '\0';
                    sscanf(opponentDirectionBuf, "{%f}", &opponentDirection);
                } else {
                    printf("Error receiving opponent direction\n");
                    gameRunning = 0;
                }
            } while (bytesRead <= 0 && gameRunning);

            // left paddle movement
            if (opponentDirection > 0.5f && leftPaddle.y > 0) {
                leftPaddle.y -= 5;
            } else if (opponentDirection < 0.5f && leftPaddle.y < height - PADDLE_HEIGHT) {
                leftPaddle.y += 5;
            }
        } else if (USER_CONTROLLED_PADDLE) {
            if (keyState[SDL_SCANCODE_UP] && leftPaddle.y > 0) leftPaddle.y -= 5;
            if (keyState[SDL_SCANCODE_DOWN] && leftPaddle.y < height - PADDLE_HEIGHT) leftPaddle.y += 5;
        } else {
            if (leftPaddle.y + PADDLE_HEIGHT/2 > ball.y+20 && leftPaddle.y > 0) {
                leftPaddle.y -= 3;
            } else if (leftPaddle.y + PADDLE_HEIGHT/2 < ball.y-20 && leftPaddle.y < height - PADDLE_HEIGHT) {
                leftPaddle.y += 3;
            } else {
                // prevent stale back and forth along a horizontal line
                if (rand() % 2 == 0 && leftPaddle.y < height - PADDLE_HEIGHT) {
                    leftPaddle.y += 3;
                } else if (leftPaddle.y > 0) {
                    leftPaddle.y -= 3;
                }
            }
        }

        // check render.txt file to see if game should render
        if (framesSinceFileRead % 60 == 0) {
            FILE *renderFd = fopen("./render.txt", "r");
            if (renderFd == NULL) {
                printf("Error: Could not read render.txt\n");
                renderGame = 0;
            } else if (fscanf(renderFd, "%d", &renderGame) != 1) {
                printf("Error: Could not read render.txt\n");
                renderGame = 0;
            }
            fclose(renderFd);
            framesSinceFileRead = 0;
        } else {
            framesSinceFileRead++;
        }

        //
        // Graphics
        //

        if (renderGame) {
            // clear screen to black
            SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255); // black background
            SDL_RenderClear(renderer);

            // draw line and point text
            SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);
            SDL_RenderFillRectF(renderer, &line);

            // draw ball and paddles
            SDL_RenderFillRectF(renderer, &ball.rect);
            SDL_RenderFillRectF(renderer, &leftPaddle.rect);
            SDL_RenderFillRectF(renderer, &rightPaddle.rect);

            // display updated frame
            SDL_RenderPresent(renderer);

            SDL_Delay(1000/FPS);
        }
    }
    
    close(clientSock1);
    close(clientSock2);
    SDL_DestroyWindow(window);
    SDL_Quit();
    return 0;
}

// sudo cp -r SDL2.framework /Library/Frameworks/
// gcc game.c -o game -Wall -F/Library/Frameworks -framework SDL2 -rpath /Library/Frameworks && ./game